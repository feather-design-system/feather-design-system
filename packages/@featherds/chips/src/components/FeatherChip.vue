<script>
import { inject, h } from "vue";
import GridChip from "./chips/GridChip";
import ReadonlyChip from "./chips/ReadonlyChip";
import RadioChip from "./chips/RadioChip";

export default {
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    labels: {
      type: Object,
    },
    condensed: {
      type: Boolean,
      default: false,
    },
    value: undefined,
  },
  setup() {
    const format = inject("chipListFormat", false);
    return { format };
  },
  render() {
    const renderChip = (component) => {
      return h(
        component,
        {
          ...this.$props,
          ...this.$attrs,
        },
        {
          default: this.$slots.default,
          ...(this.$slots.icon && { icon: this.$slots.icon }),
        }
      );
    };
    if (this.format === "grid") {
      return renderChip(GridChip);
    }
    if (this.format === "radio") {
      return renderChip(RadioChip);
    }
    return renderChip(ReadonlyChip);
  },
};
</script>
