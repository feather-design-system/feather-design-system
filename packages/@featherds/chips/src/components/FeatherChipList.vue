<template>
  <div
    v-bind="attrs"
    data-ref-id="feather-chip-list"
    :aria-label="label"
    class="chip-list"
    :class="{ condensed: condensed, single: single }"
    @keydown="keydown"
  >
    <slot />
  </div>
</template>
<script lang="ts">
import { provide, toRef, defineComponent, ref } from "vue";
import { useRadioGroup } from "@featherds/composables/radio/RadioGroup";
export const props = {
  label: {
    type: String,
    required: true,
  },
  mode: {
    type: String,
    default: "list",
    validator: (v: string) => {
      // The value must match either
      return ["list", "radio", "single"].indexOf(v) !== -1;
    },
  },
  modelValue: {
    type: [String, Number, Boolean, Array, Object, Date, Function],
  },
  condensed: {
    type: Boolean,
    default: false,
  },
};
export default defineComponent({
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  props,
  setup(props, context) {
    const format = props.mode === "list" ? "grid" : props.mode;
    provide("chipListFormat", format);
    const single = format === "single";

    if (format === "radio") {
      const modelValue = toRef(props, "modelValue");
      const attrs = {
        role: "radiogroup",
      };
      return {
        attrs,
        ...useRadioGroup(
          modelValue,
          context.emit,
          props.label as string,
          {},
          ref("")
        ),
        single,
      };
    }
    // if (format === "grid" || format === "single") {
    const attrs = {
      role: "grid",
    };
    const keydown = () => {};

    return { attrs, keydown, single };
  },
});
</script>

<style lang="scss" scoped>
@import "../../scss/mixins";
.chip-list {
  margin: 0;
  margin-top: 0.75rem;
  margin-bottom: 0.125rem;
  padding: 0;
  list-style: none;
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-wrap: wrap;

  & + .chip-list {
    margin-top: 1.75rem;
  }
  &.condensed {
    :deep(.chip) {
      @include condensed-chip;
    }
  }
  &.single {
    margin: 0;
    :deep(.chip) {
      @include standalone-chip;
    }
  }
}
</style>
