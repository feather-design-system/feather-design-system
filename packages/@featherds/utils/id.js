const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
