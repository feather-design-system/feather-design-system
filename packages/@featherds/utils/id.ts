const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
