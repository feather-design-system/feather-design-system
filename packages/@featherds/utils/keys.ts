const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
