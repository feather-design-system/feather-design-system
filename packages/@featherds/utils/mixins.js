var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
