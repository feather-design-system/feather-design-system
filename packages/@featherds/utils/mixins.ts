const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
