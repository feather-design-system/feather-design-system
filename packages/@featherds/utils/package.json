{
  "name": "@featherds/utils",
  "version": "0.2.10",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {},
  "keywords": [
    "vue",
    "FeatherUtils"
  ],
  "types": "./src/index.d.ts"
}
