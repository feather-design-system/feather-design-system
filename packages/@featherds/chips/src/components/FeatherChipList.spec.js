import FeatherChipList from "./FeatherChipList.vue";
import FeatherChip from "./FeatherChip.vue";
import { mount } from "@vue/test-utils";
import * as id from "@featherds/utils/id";
import { h, nextTick } from "vue";
jest.spyOn(id, "getSafeId").mockImplementation((x) => x);

import axe from "@featherds/utils/test/axe";

const getRadio = (value, disabled) => getChip(disabled, false, false, value);
const getChip = (disabled = false, del = false, click = false, value) => {
  const test = () => {
    return;
  };
  const on = {};
  if (del) {
    on.onDelete = test;
  }
  if (click) {
    on.onClick = test;
  }
  return h(
    FeatherChip,
    {
      disabled: disabled,
      value: value,
      ...on,
    },
    {
      default: () => "TEST",
    }
  );
};

const getWrapper = (chips, options = {}) => {
  options.props = {
    ...(options.props || {}),
    label: "TEST",
  };
  options.slots = {
    default: () => chips,
  };
  return mount(FeatherChipList, options);
};

describe("Feather Chip List", () => {
  describe("list format", () => {
    it("should render child chips correctly", () => {
      const chips = [getChip(), getChip()];
      const wrapper = getWrapper(chips);
      expect(wrapper.wrapperElement).toMatchSnapshot();
    });
    it("should render child chips with delete correctly", () => {
      const chips = [getChip(false, true), getChip(false, true)];
      const wrapper = getWrapper(chips);
      expect(wrapper.wrapperElement).toMatchSnapshot();
    });

    it("should render child chips with click correctly", () => {
      const chips = [getChip(false, false, true), getChip(false, false, true)];
      const wrapper = getWrapper(chips);
      expect(wrapper.wrapperElement).toMatchSnapshot();
    });
    it("should render child chips with one disabled correctly", () => {
      const chips = [getChip(false), getChip(true)];
      const wrapper = getWrapper(chips);
      expect(wrapper.wrapperElement).toMatchSnapshot();
    });

    it("should be accessible", async () => {
      const chips = [getChip()];
      const wrapper = getWrapper(chips);
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
    it("should be accessible with click", async () => {
      const chips = [getChip(false, false, true)];
      const wrapper = getWrapper(chips);
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
    it("should be accessible with delete", async () => {
      const chips = [getChip(false, true)];
      const wrapper = getWrapper(chips);
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
    it("should be accessible with one disabled", async () => {
      const chips = [getChip(true)];
      const wrapper = getWrapper(chips);
      expect(await axe(wrapper.element)).toHaveNoViolations();
    });
  });
  describe("radio format", () => {
    const selector = "[role='radiogroup']";
    it("should render child chips to correctly", async () => {
      const wrapper = getWrapper([getRadio(1, false), getRadio(2, false)], {
        props: { mode: "radio", modelValue: 1 },
      });
      await nextTick();
      await nextTick();

      expect(wrapper.wrapperElement).toMatchSnapshot();
    });
    it("should set first property on first radio if none match the value", async () => {
      const wrapper = getWrapper([getRadio(1, false), getRadio(2, false)], {
        props: { mode: "radio", modelValue: -1 },
      });
      await nextTick();
      await nextTick();
      expect(wrapper.findAll("[role='radio']")[0].attributes("tabindex")).toBe(
        "0"
      );
    });
    it("should set first property on first radio not disabled if none match the value", async () => {
      const wrapper = getWrapper([getRadio(1, true), getRadio(2, false)], {
        props: { mode: "radio", modelValue: -1 },
      });
      await nextTick();
      await nextTick();
      expect(wrapper.findAll("[role='radio']")[1].attributes("tabindex")).toBe(
        "0"
      );
    });
    it("should set select first radio that matches initial value", async () => {
      const wrapper = getWrapper([getRadio(1, false), getRadio(2, false)], {
        props: { mode: "radio", modelValue: 2 },
      });
      await nextTick();
      await nextTick();
      await nextTick();
      expect(
        wrapper.findAll("[role='radio']")[1].attributes("aria-checked")
      ).toBe("true");
    });
    it("should emit input on check", async () => {
      const wrapper = getWrapper([getRadio(1, false), getRadio(2, false)], {
        props: { mode: "radio", modelValue: 2 },
      });
      await wrapper.find("[role='radio']").trigger("click");
      expect(wrapper.emitted("update:modelValue")[0][0]).toBe(1);
    });

    it("should select clicked radio button", async () => {
      const wrapper = getWrapper([getRadio(1, false), getRadio(2, false)], {
        props: { mode: "radio", modelValue: -1 },
      });
      await nextTick();
      await nextTick();
      await wrapper.find("[role='radio']").trigger("click");
      expect(wrapper.emitted("update:modelValue")[0][0]).toBe(1);
    });
    it("should not select a disabled clicked radio button", async () => {
      const wrapper = getWrapper([getRadio(1, true), getRadio(2, false)], {
        props: { mode: "radio", modelValue: -1 },
      });
      await nextTick();
      await nextTick();
      await wrapper.find("[role='radio']").trigger("click");
      expect(wrapper.emitted("update:modelValue")).toBeUndefined();
    });
    it("should select current radio when space or enter is pressed", async () => {
      const testSelection = async (code) => {
        const wrapper = getWrapper([getRadio(1, false), getRadio(2, false)], {
          props: { mode: "radio", modelValue: -1 },
        });
        await nextTick();
        await nextTick();
        await wrapper.find(selector).trigger(`keydown.${code}`);
        expect(wrapper.emitted("update:modelValue")[0][0]).toBe(1);
      };
      await testSelection("enter");
      await testSelection("space");
    });
    it("should select next enabled radio when right or down is pressed", async () => {
      const testNext = async (code) => {
        const wrapper = getWrapper(
          [getRadio(1, false), getRadio(2, true), getRadio(3, false)],
          {
            props: { mode: "radio", modelValue: 1 },
          }
        );
        await nextTick();
        await nextTick();
        await wrapper.find(selector).trigger(`keydown.${code}`);
        expect(wrapper.emitted("update:modelValue")[0][0]).toBe(3);
      };
      await testNext("down");
      await testNext("right");
    });
    it("should select first enabled radio when right or down is pressed at the end of the group", async () => {
      const testNext = async (code) => {
        const wrapper = getWrapper(
          [getRadio(1, true), getRadio(2, false), getRadio(3, false)],
          {
            props: { mode: "radio", modelValue: 3 },
          }
        );
        await nextTick();
        await nextTick();
        await wrapper.find(selector).trigger(`keydown.${code}`);
        expect(wrapper.emitted("update:modelValue")[0][0]).toBe(2);
      };
      await testNext("down");
      await testNext("right");
    });
    it("should select previous enabled radio when left or up is pressed", async () => {
      const testPrev = async (code) => {
        const wrapper = getWrapper(
          [getRadio(1, false), getRadio(2, true), getRadio(3, false)],
          {
            props: { mode: "radio", modelValue: 3 },
          }
        );

        await nextTick();
        await nextTick();
        await wrapper.find(selector).trigger(`keydown.${code}`);
        expect(wrapper.emitted("update:modelValue")[0][0]).toBe(1);
      };
      await testPrev("left");
      await testPrev("up");
    });
    it("should select last enabled radio when left or up is pressed at the start of the group", async () => {
      const testPrev = async (code) => {
        const wrapper = getWrapper(
          [
            getRadio(1, false),
            getRadio(0, false),
            getRadio(2, false),
            getRadio(3, true),
          ],
          {
            props: { mode: "radio", modelValue: 1 },
          }
        );

        await nextTick();
        await nextTick();
        await wrapper.find(selector).trigger(`keydown.${code}`);
        expect(wrapper.emitted("update:modelValue")[0][0]).toBe(2);
      };
      await testPrev("left");
      await testPrev("up");
    });
  });
});
