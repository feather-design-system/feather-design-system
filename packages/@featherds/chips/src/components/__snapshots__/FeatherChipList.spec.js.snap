// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feather Chip List list format should render child chips correctly 1`] = `
<div
  aria-label="TEST"
  class="chip-list"
  data-ref-id="feather-chip-list"
  role="grid"
>
  
  <div
    aria-disabled="false"
    class="chip"
    role="row"
  >
    
    <span
      aria-disabled="false"
      role="gridcell"
    >
      <span
        class="chip-label-button"
      >
        <!--v-if-->
        <span
          class="label"
          id="chip-text"
          title="false"
        >
          
          
          TEST
          
          
        </span>
      </span>
    </span>
    <!--v-if-->
    
    <!--v-if-->
  </div>
  <div
    aria-disabled="false"
    class="chip"
    role="row"
  >
    
    <span
      aria-disabled="false"
      role="gridcell"
    >
      <span
        class="chip-label-button"
      >
        <!--v-if-->
        <span
          class="label"
          id="chip-text"
          title="false"
        >
          
          
          TEST
          
          
        </span>
      </span>
    </span>
    <!--v-if-->
    
    <!--v-if-->
  </div>
  
</div>
`;

exports[`Feather Chip List list format should render child chips with click correctly 1`] = `
<div
  aria-label="TEST"
  class="chip-list"
  data-ref-id="feather-chip-list"
  role="grid"
>
  
  <div
    aria-disabled="false"
    class="chip hover focus"
    role="row"
    style="overflow: hidden; position: relative;"
  >
    
    <span
      aria-disabled="false"
      role="gridcell"
    >
      <span
        class="chip-label-button"
        role="button"
        tabindex="0"
      >
        <!--v-if-->
        <span
          class="label"
          id="chip-text"
          title="false"
        >
          
          
          TEST
          
          
        </span>
      </span>
    </span>
    <!--v-if-->
    
    <!---->
  </div>
  <div
    aria-disabled="false"
    class="chip hover focus"
    role="row"
    style="overflow: hidden; position: relative;"
  >
    
    <span
      aria-disabled="false"
      role="gridcell"
    >
      <span
        class="chip-label-button"
        role="button"
        tabindex="0"
      >
        <!--v-if-->
        <span
          class="label"
          id="chip-text"
          title="false"
        >
          
          
          TEST
          
          
        </span>
      </span>
    </span>
    <!--v-if-->
    
    <!---->
  </div>
  
</div>
`;

exports[`Feather Chip List list format should render child chips with delete correctly 1`] = `
<div
  aria-label="TEST"
  class="chip-list"
  data-ref-id="feather-chip-list"
  role="grid"
>
  
  <div
    aria-disabled="false"
    class="chip focus"
    role="row"
  >
    
    <span
      aria-disabled="false"
      role="gridcell"
    >
      <span
        class="chip-label-button"
      >
        <!--v-if-->
        <span
          class="label"
          id="chip-text"
          title="false"
        >
          
          
          TEST
          
          
        </span>
      </span>
    </span>
    <span
      class="chip-delete"
      role="gridcell"
    >
      <a
        aria-describedby="chip-text"
        aria-label="Remove"
        class="delete-icon"
        href="#"
        role="button"
      >
        <svg
          aria-hidden="true"
          class="feather-icon feather-icon-flex"
          data-v-0aa06a12=""
          focusable="false"
          height="24"
          role="img"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12,2A10,10,0,1,0,22,12,10,10,0,0,0,12,2Zm5,13.24A1.26,1.26,0,0,1,17,17a1.24,1.24,0,0,1-1.76,0L12,13.73,8.76,17A1.24,1.24,0,1,1,7,15.24L10.24,12,7,8.76A1.24,1.24,0,0,1,8.76,7L12,10.2,15.24,7A1.24,1.24,0,1,1,17,8.76L13.76,12Z"
          />
        </svg>
      </a>
    </span>
    
    <!--v-if-->
  </div>
  <div
    aria-disabled="false"
    class="chip focus"
    role="row"
  >
    
    <span
      aria-disabled="false"
      role="gridcell"
    >
      <span
        class="chip-label-button"
      >
        <!--v-if-->
        <span
          class="label"
          id="chip-text"
          title="false"
        >
          
          
          TEST
          
          
        </span>
      </span>
    </span>
    <span
      class="chip-delete"
      role="gridcell"
    >
      <a
        aria-describedby="chip-text"
        aria-label="Remove"
        class="delete-icon"
        href="#"
        role="button"
      >
        <svg
          aria-hidden="true"
          class="feather-icon feather-icon-flex"
          data-v-0aa06a12=""
          focusable="false"
          height="24"
          role="img"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12,2A10,10,0,1,0,22,12,10,10,0,0,0,12,2Zm5,13.24A1.26,1.26,0,0,1,17,17a1.24,1.24,0,0,1-1.76,0L12,13.73,8.76,17A1.24,1.24,0,1,1,7,15.24L10.24,12,7,8.76A1.24,1.24,0,0,1,8.76,7L12,10.2,15.24,7A1.24,1.24,0,1,1,17,8.76L13.76,12Z"
          />
        </svg>
      </a>
    </span>
    
    <!--v-if-->
  </div>
  
</div>
`;

exports[`Feather Chip List list format should render child chips with one disabled correctly 1`] = `
<div
  aria-label="TEST"
  class="chip-list"
  data-ref-id="feather-chip-list"
  role="grid"
>
  
  <div
    aria-disabled="false"
    class="chip"
    role="row"
  >
    
    <span
      aria-disabled="false"
      role="gridcell"
    >
      <span
        class="chip-label-button"
      >
        <!--v-if-->
        <span
          class="label"
          id="chip-text"
          title="false"
        >
          
          
          TEST
          
          
        </span>
      </span>
    </span>
    <!--v-if-->
    
    <!--v-if-->
  </div>
  <div
    aria-disabled="true"
    class="chip disabled"
    role="row"
  >
    
    <span
      aria-disabled="true"
      role="gridcell"
    >
      <span
        class="chip-label-button"
      >
        <!--v-if-->
        <span
          class="label"
          id="chip-text"
          title="false"
        >
          
          
          TEST
          
          
        </span>
      </span>
    </span>
    <!--v-if-->
    
    <!--v-if-->
  </div>
  
</div>
`;

exports[`Feather Chip List radio format should render child chips to correctly 1`] = `
<div
  aria-label="TEST"
  class="chip-list"
  data-ref-id="feather-chip-list"
  role="radiogroup"
>
  
  <div
    allow-click="true"
    aria-checked="true"
    aria-disabled="false"
    aria-labelledby="chip-label-id"
    class="chip focus hover selected"
    role="radio"
    tabindex="0"
  >
    
    <!--v-if-->
    <span
      class="label"
      id="chip-label-id"
      title="false"
    >
      
      
      TEST
      
      
    </span>
    
    <!--v-if-->
  </div>
  <div
    allow-click="true"
    aria-checked="false"
    aria-disabled="false"
    aria-labelledby="chip-label-id"
    class="chip focus hover"
    role="radio"
    tabindex="-1"
  >
    
    <!--v-if-->
    <span
      class="label"
      id="chip-label-id"
      title="false"
    >
      
      
      TEST
      
      
    </span>
    
    <!--v-if-->
  </div>
  
</div>
`;
